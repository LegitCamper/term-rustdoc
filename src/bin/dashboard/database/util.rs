use super::Progress;
use crate::{dashboard::database::cache_info::CachedDocInfo, local_registry::PkgInfo, Result};
use bincode::config;
use serde::{de::DeserializeOwned, Serialize};
use std::path::PathBuf;

pub fn build(progress: Progress, db_dir: PathBuf, pkg_dir: PathBuf, pkg_info: PkgInfo) {
    let mut cargo_toml = pkg_dir;
    cargo_toml.push("Cargo.toml");
    rayon::spawn(move || {
        let dir = match tempfile::tempdir() {
            Ok(dir) => dir,
            Err(err) => {
                error!("Can't create a tempdir:\n{err}");
                return;
            }
        };
        let mut cache_info = CachedDocInfo::new(pkg_info.to_name_ver(), db_dir);
        info!(?cache_info.pkg, "begin to compile the doc under {}", dir.path().display());
        match rustdoc_json::Builder::default()
            .toolchain("nightly")
            .silent(true)
            .target_dir(&dir)
            .manifest_path(&cargo_toml)
            .build()
        {
            Ok(json_path) => {
                let meta = cache_info.meta_mut();
                meta.set_finished_duration();
                let duration = meta.duration_as_secs();
                info!(?cache_info.pkg, ?json_path, "succeefully compiled the doc in {duration:.2}s");
                if let Err(err) = cache_info.save_doc(&json_path, pkg_info) {
                    error!("{err}");
                }
                match progress.lock() {
                    Ok(mut v) => v.push(cache_info),
                    Err(err) => {
                        error!(
                            "Failed to lock the progress to write generated PkgKey.\
                                 The doc is generated though.\n{err}"
                        )
                    }
                }
            }
            Err(err) => error!("Failed to compile {}:\n{err}", cargo_toml.display()),
        }
    });
}

pub fn encode<T: Serialize>(t: T) -> Result<Vec<u8>> {
    Ok(bincode::serde::encode_to_vec(t, config::standard())?)
}

pub fn decode<T: DeserializeOwned>(v: &[u8]) -> Result<T> {
    Ok(bincode::serde::decode_from_slice(v, config::standard())?.0)
}
