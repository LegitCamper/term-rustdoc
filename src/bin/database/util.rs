use super::PkgKey;
use crate::{
    database::CachedDocInfo,
    event::{Event, Sender},
    local_registry::PkgInfo,
    Result,
};
use bincode::config;
use bytesize::ByteSize;
use serde::{de::DeserializeOwned, Serialize};
use std::{io::Write, path::PathBuf};
use xz2::write::{XzDecoder, XzEncoder};

pub fn build(sender: Sender, db_dir: PathBuf, pkg_dir: PathBuf, pkg_info: PkgInfo) -> PkgKey {
    let mut cargo_toml = pkg_dir;
    cargo_toml.push("Cargo.toml");
    let in_progress = PkgKey::new_with_default_feature(pkg_info.to_name_ver());
    rayon::spawn(move || {
        let dir = match tempfile::tempdir() {
            Ok(dir) => dir,
            Err(err) => {
                error!("Can't create a tempdir:\n{err}");
                return;
            }
        };
        let mut cache_info =
            CachedDocInfo::new_with_default_feature(pkg_info.to_name_ver(), db_dir);
        info!(?cache_info.pkg, "begin to compile the doc under {}", dir.path().display());
        match rustdoc_json::Builder::default()
            .toolchain("nightly")
            .silent(true)
            .target_dir(&dir)
            .manifest_path(&cargo_toml)
            .build()
        {
            Ok(json_path) => {
                let meta = cache_info.meta_mut();
                meta.set_finished_duration();
                let duration = meta.duration_as_secs();
                info!(?cache_info.pkg, ?json_path, "succeefully compiled the doc in {duration:.2}s");
                if let Err(err) = cache_info.save_doc(&json_path, pkg_info) {
                    error!("{err}");
                }
                match sender.send(Event::DocCompiled(Box::new(cache_info))) {
                    Ok(()) => (),
                    Err(err) => {
                        error!(
                            "Failed to send `DocCompiled` event when CachedDocInfo is ready:\n{err}"
                        )
                    }
                }
            }
            Err(err) => error!("Failed to compile {}:\n{err}", cargo_toml.display()),
        }
    });
    in_progress
}

/// Write source data into db file.
pub fn encode<T: Serialize>(t: T) -> Result<Vec<u8>> {
    Ok(bincode::serde::encode_to_vec(t, config::standard())?)
}

/// Read data from db file.
pub fn decode<T: DeserializeOwned>(raw: &[u8]) -> Result<T> {
    Ok(bincode::serde::decode_from_slice(raw, config::standard())?.0)
}

/// Write source data into db file with xz compression.
///
/// NOTE: not all bytes are compressed via xz, because compression on small bytes
/// will increase the size.
/// For now, only compress the doc data, i.e. host-json and host-parsed table.
/// Be careful to decompress them before deserialization.
pub fn encode_with_xz<T: Serialize>(t: T) -> Result<Vec<u8>> {
    let raw = bincode::serde::encode_to_vec(t, config::standard())?;
    xz_encode_on_bytes(&raw)
}

pub fn xz_encode_on_bytes(raw: &[u8]) -> Result<Vec<u8>> {
    let mut compressed = Vec::with_capacity(raw.len() / 2);
    let mut xz_encoder = XzEncoder::new(&mut compressed, 9);
    xz_encoder.write_all(raw)?;
    xz_encoder.finish()?;
    compressed.shrink_to_fit();
    let (before, after) = (raw.len(), compressed.len());
    info!(
        "compress {} => {} (reduced {:.1}%)",
        ByteSize(before as u64),
        ByteSize(after as u64),
        (after as f32 / before as f32 - 1.0) * 100.0
    );
    Ok(compressed)
}

/// Read data from db file with xz decompression.
pub fn decode_with_xz<T: DeserializeOwned>(raw: &[u8]) -> Result<T> {
    let decompressed = xz_decode_on_bytes(raw)?;
    Ok(bincode::serde::decode_from_slice(&decompressed, config::standard())?.0)
}

pub fn xz_decode_on_bytes(raw: &[u8]) -> Result<Vec<u8>> {
    let mut decompressed = Vec::with_capacity(raw.len() * 4);
    {
        let mut xz_decoder = XzDecoder::new(&mut decompressed);
        xz_decoder.write_all(raw)?;
        xz_decoder.finish()?;
    }
    let (before, after) = (raw.len(), decompressed.len());
    info!(
        "decompress {} => {} (ratio {:.1}%)",
        ByteSize(before as u64),
        ByteSize(after as u64),
        (1.0 - before as f32 / after as f32) * 100.0
    );
    Ok(decompressed)
}
