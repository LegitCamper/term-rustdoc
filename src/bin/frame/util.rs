use ratatui::{
    layout::Flex,
    prelude::{Constraint, Layout, Rect},
};

pub fn centered_rect(area: Rect, width: u16, height: u16) -> Rect {
    let horizontal = Layout::horizontal([Constraint::Percentage(width)]).flex(Flex::Center);
    let vertical = Layout::vertical([Constraint::Percentage(height)]).flex(Flex::Center);
    let [area] = vertical.areas(area);
    let [area] = horizontal.areas(area);
    area
}
