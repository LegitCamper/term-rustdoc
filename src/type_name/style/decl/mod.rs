mod function;
mod struct_;

use super::{
    path::{FindName, Format, Short},
    StyledType, Vis,
};
use crate::tree::IDMap;
use rustdoc_types::{ItemEnum, Visibility};

pub fn item_styled(id: &str, map: &IDMap) -> StyledType {
    if let Some(item) = map.get_item(id) {
        let vis_name_map = VisNameMap {
            name: item.name.as_deref().unwrap_or(""),
            vis: &item.visibility,
            map,
        };
        let mut buf = StyledType::with_capacity(48);
        match &item.inner {
            ItemEnum::Import(reexport) => {
                let id = reexport.id.as_ref();
                return id.map(|id| item_styled(&id.0, map)).unwrap_or_default();
            }
            ItemEnum::Function(f) => f.format_as_short(vis_name_map, &mut buf),
            ItemEnum::Struct(s) => s.format_as_short(vis_name_map, &mut buf),
            _ => return StyledType::default(),
        };
        return buf;
    }
    StyledType::default()
}

impl Format for Visibility {
    fn format<Kind: FindName>(&self, buf: &mut StyledType) {
        buf.write(match self {
            Visibility::Public => Vis::Pub,
            Visibility::Default => Vis::Default,
            Visibility::Crate => Vis::PubCrate,
            Visibility::Restricted { parent, path } => {
                buf.write_vis_scope(parent, path);
                return;
            }
        });
    }
}

struct VisNameMap<'a> {
    vis: &'a Visibility,
    name: &'a str,
    map: &'a IDMap,
}

trait Declaration {
    fn format<K: FindName>(&self, map: VisNameMap, buf: &mut StyledType);
    fn format_as_short(&self, map: VisNameMap, buf: &mut StyledType) {
        <Self as Declaration>::format::<Short>(self, map, buf);
    }
}
