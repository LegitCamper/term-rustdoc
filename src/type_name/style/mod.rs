#![allow(unused)]
mod path;
mod type_;

use crate::{
    tree::{IDMap, IdToID, ID},
    util::XString,
};

use self::path::{FindName, Format};

pub struct StyledType {
    inner: Vec<Tag>,
}

impl StyledType {
    fn write<T: Into<Tag>>(&mut self, tag: T) {
        self.inner.push(tag.into());
    }

    fn write_format<Kind: FindName>(&mut self, fmt: impl Format) {
        fmt.format::<Kind>(self);
    }

    #[allow(clippy::inherent_to_string)]
    fn to_string(&self) -> String {
        let cap = self.inner.iter().map(Tag::str_len).sum::<usize>();
        let mut buf = String::with_capacity(cap);
        for tag in &self.inner {
            tag.write_to_string(&mut buf)
        }
        buf
    }

    fn write_name(&mut self, name: &str) {
        self.write(Tag::Name(name.into()));
    }

    fn write_id_name(&mut self, id: impl IdToID, name: &str) {
        self.write(Tag::Path(id.to_ID()));
        self.write(Tag::Name(name.into()));
    }

    fn write_vis_scope(&mut self, id: ID, map: &IDMap) {
        self.write(Tag::Decl(Decl::Vis(Vis::PubScope)));
        self.write_in_parentheses(|s| {
            let name = map.path(&id);
            s.write(Tag::PubScope(id));
            s.write(Tag::Name(name))
        });
    }

    /// Write `start_tag` `x` `end_tag` where x is written from the callback.
    /// (start_tag, end_tag) can be `()` `[]` `{}` `<>` [`Span`] etc.
    fn write_enclosing_tag(
        &mut self,
        start_tag: Tag,
        end_tag: Tag,
        f: impl FnOnce(&mut StyledType),
    ) {
        self.write(start_tag);
        f(self);
        self.write(end_tag);
    }
}

macro_rules! impl_write_tag {
    (@Punctuation $($fname:ident $s:literal: $start:ident, $end:ident),+ $(,)?) => {
        /// Write stuff in enclosing [`Punctuation`]s.
        impl StyledType { $(
            #[doc = concat!("Write `", $s, "` where x is written from the callback.")]
            pub fn $fname(&mut self, f: impl FnOnce(&mut StyledType)) {
                let start = Tag::Symbol(Symbol::Punctuation(Punctuation::$start));
                let end = Tag::Symbol(Symbol::Punctuation(Punctuation::$end));
                self.write_enclosing_tag(start, end, f);
            }
        )+ }
    };
    (@span $($fname:ident $s:literal $span:ident),+ $(,)?) => {
        /// Write stuff in enclosing [`Span`]s.
        impl StyledType { $(
            #[doc = concat!("Write `", $s, "` as [`Span::", stringify!($span),
              "`] between [`Tag::Start`] and [`Tag::End`] with the callback.")]
            pub fn $fname(&mut self, f: impl FnOnce(&mut StyledType)) {
                let start = Tag::Start(Span::$span);
                let end = Tag::End(Span::$span);
                self.write_enclosing_tag(start, end, f);
            }
        )+ }
    };
}

impl_write_tag!(@Punctuation
    write_in_brace "{x}": BraceStart, BraceEnd ,
    write_in_parentheses "(x)": ParenthesisStart, ParenthesisEnd ,
    write_in_angle_bracket "<x>": AngleBracketStart, AngleBracketEnd ,
    write_in_squre_bracket "[x]": SquareBracketStart, SquareBracketEnd ,
);

impl_write_tag!(@span
    write_span_path_name "path_without_generics" PathName ,
    write_span_type_name "..." TypeName ,
    write_span_where_bound "where ..." WhereBound ,
    write_span_generics_def "< ... >" GenericsDef ,
    write_span_function_name "fn ..." FunctionName ,
);

/// Rendering tag which represents color, style and metadata that are used to jump.
#[derive(Clone, Debug)]
pub enum Tag {
    /// A path to an item that is usually carries an ID.
    /// We use the ID to jump to another item.
    /// The Path do not include generics.
    /// A Path ID tag is conjuction with its Name tag.
    Path(ID),
    /// A Name is a short path an ID can refer to or somthing not statically known.
    /// E.g. short path/type name, function argument, name for field, variant and generics etc.
    Name(XString),
    Symbol(Symbol),
    Decl(Decl),
    /// The scope id is in conjuction with a Name.
    /// `pub(scope)` is composed of [`Vis::PubScope`], [`Punctuation::ParenthesisStart`],
    /// [`Tag::Name`] and [`Punctuation::ParenthesisEnd`].
    PubScope(ID),
    /// In conjuction with [`Abi`].
    /// `extern "other_abi" ` is composed of [`Abi::Other`], [`Punctuation::Quote`],
    /// [`Tag::UnusualAbi`], [`Punctuation::Quote`] and [`Punctuation::WhiteSpace`].
    UnusualAbi(XString),
    Start(Span),
    End(Span),
}

impl Tag {
    pub fn write_to_string(&self, buf: &mut String) {
        match self {
            Tag::Path(_) => (),
            Tag::Name(s) => buf.push_str(s),
            Tag::Symbol(s) => buf.push_str(s.to_str()),
            Tag::Decl(s) => buf.push_str(s.to_str()),
            Tag::PubScope(_) => (),
            Tag::UnusualAbi(s) => buf.push_str(s),
            Tag::Start(_) | Tag::End(_) => (),
        }
    }
    pub fn str_len(&self) -> usize {
        match self {
            // a path uses name len instead of id len
            Tag::Path(_) => 0,
            Tag::Name(s) => s.len(),
            Tag::Symbol(s) => s.str_len(),
            Tag::Decl(s) => s.str_len(),
            // `pub(scope)`: scope use name len rather than id len, but we should count `pub()` here
            Tag::PubScope(_) => 0,
            Tag::UnusualAbi(s) => s.len(),
            Tag::Start(_) | Tag::End(_) => 0,
        }
    }
}

// macro MWE for interspersing fieldless variants with field-carrying ones:
// https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=33ef54aa78ae2ac682c911440f8d576a

/// Implement to_str and str_len methods and basic Derive macros for a fieldless enum.
macro_rules! to_str {
    ($({$val:ident $from:expr})?
        $(#[$em:meta])*
        $vis:vis enum $e:ident { $($t:tt)+ }
    ) => {
        to_str!(@impl [def {$(#[$em])* $vis} $e {}] [to_str {}] [str_len {}] : $($t)+);
        $(impl From<$e> for Tag {
            fn from($val: $e) -> Tag { $from }
        })?
    };
    // expand token trees
    (@impl
     [def {$(#[$em:meta])* $vis:vis} $e:ident { $($vars:tt)* }]
     [to_str  { $($b1:tt)* }]
     [str_len { $($b2:tt)* }] :
    ) => {
        #[derive(Clone, Copy, Debug)] $(#[$em])*
        $vis enum $e { $($vars)* }
        impl $e {
            pub const fn to_str(self) -> &'static str {
                match self { $($b1)* }
            }
            pub const fn str_len(self) -> usize {
                match self { $($b2)* }
            }
        }
    };
    (@impl
     [def {$(#[$em:meta])* $vis:vis} $e:ident { $($vars:tt)* }]
     [to_str  { $($b1:tt)* }]
     [str_len { $($b2:tt)* }] :
     $(#[$vm:meta])* $var:ident = $s:literal ,
    ) => {
        to_str!(@impl
            [def {$(#[$em])* $vis} $e { $($vars)* $(#[$vm])* $var , } ]
            [to_str  { $($b1)* $e::$var => $s,       } ]
            [str_len { $($b2)* $e::$var => $s.len(), } ] :
        );
    };
    (@impl
     [def {$(#[$em:meta])* $vis:vis} $e:ident { $($vars:tt)* }]
     [to_str  { $($b1:tt)* }]
     [str_len { $($b2:tt)* }] :
     $(#[$vm:meta])* $var:ident = $s:literal , $($t:tt)+
    ) => {
        to_str!(@impl
            [def {$(#[$em])* $vis} $e { $($vars)* $(#[$vm])* $var , } ]
            [to_str  { $($b1)* $e::$var => $s,       } ]
            [str_len { $($b2)* $e::$var => $s.len(), } ] :
            $($t)+
        );
    };
    (@impl
     [def {$(#[$em:meta])* $vis:vis} $e:ident { $($vars:tt)* }]
     [to_str  { $($b1:tt)* }]
     [str_len { $($b2:tt)* }] :
     $(#[$vm:meta])* $var:ident($inner:ident) ,
    ) => {
        to_str!(@impl
            [def {$(#[$em])* $vis} $e { $($vars)* $(#[$vm])* $var($inner) , } ]
            [to_str  { $($b1)* $e::$var(val) => val.to_str(),  } ]
            [str_len { $($b2)* $e::$var(val) => val.str_len(), } ] :
        );
    };
    (@impl
     [def {$(#[$em:meta])* $vis:vis} $e:ident { $($vars:tt)* }]
     [to_str  { $($b1:tt)* }]
     [str_len { $($b2:tt)* }] :
     $(#[$vm:meta])* $var:ident($inner:ident) , $($t:tt)+
    ) => {
        to_str!(@impl
            [def {$(#[$em])* $vis} $e { $($vars)* $(#[$vm])* $var($inner) , } ]
            [to_str  { $($b1)* $e::$var(val) => val.to_str(),  } ]
            [str_len { $($b2)* $e::$var(val) => val.str_len(), } ] :
            $($t)+
        );
    };
}

to_str!({val Tag::Symbol(val)}
    pub enum Symbol {
        Syntax(Syntax),
        Punctuation(Punctuation),
    }
);

to_str!({val Tag::Symbol(Symbol::Syntax(val))}
    /// Symbol as syntax component.
    ///
    /// NOTE: some syntax has already included whitespaces, because this saves pushing them.
    pub enum Syntax {
        Reference = "&",
        ReferenceMut = "&mut",
        /// lifetime may lie between `&` and `mut`
        Mut = "mut",
        Self_ = "Self",
        Where = "where ",
        Dyn = "dyn ",
        PathSep = "::",
        As = " as ",
        RawPointer = "*const ",
        RawPointerMut = "*mut ",
        Infer = "_",
        Impl = "impl ",
        For = "for",
        /// mainly for `?Sized`
        Maybe = "?",
        MaybeConst = "~const",
    }
);

to_str!({val Tag::Symbol(Symbol::Punctuation(val))}
    /// Punctuation symbol.
    ///
    /// NOTE: some Punctuations have included whitespaces for convenience.
    pub enum Punctuation {
        WhiteSpace = " ",
        NewLine = "\n",
        /// `, `
        Comma = ", ",
        /// `: `
        Colon = ": ",
        /// <code> = </code>
        Equal = " = ",
        /// <code> + </code>
        Plus = " + ",
        Apostrophe = "'",
        /// `; `
        SimiColon = "; ",
        AngleBracketStart = "<",
        AngleBracketEnd = ">",
        SquareBracketStart = "[",
        SquareBracketEnd = "]",
        ParenthesisStart = "(",
        ParenthesisEnd = ")",
        BraceStart = "{",
        BraceEnd = "}",
        Quote = "\"",
    }
);

to_str!({val Tag::Decl(val)}
    /// Components in declaration. A type doesn't need this, but type declaration need this.
    pub enum Decl {
        Vis(Vis),
        Function(Function),
        Struct(Struct),
    }
);

to_str!({val Tag::Decl(Decl::Vis(val))}
    pub enum Vis {
        Pub = "pub ",
        /// placeholder: not showing anything
        Default = "",
        PubCrate = "pub(crate) ",
        /// In conjuction with [`Tag::PubScope`].
        PubScope = "pub",
    }
);

to_str!({val Tag::Decl(Decl::Function(val))}
    /// FunctionQualifiers order: const? async? unsafe? (extern Abi?)? fn
    pub enum Function {
        Const = "const ",
        Async = "async ",
        Unsafe = "unsafe ",
        Abi(Abi),
        Fn = "fn ",
    }
);

to_str!({val Tag::Decl(Decl::Function(Function::Abi(val)))}
    pub enum Abi {
        /// `extern "Rust"` is valid though, but for simplicity, no need to show it.
        Rust = "",
        C = "extern \"C\" ",
        Cdecl = "extern \"cdecl\" ",
        Stdcall = "extern \"stdcall\" ",
        Fastcall = "extern \"fastcall\" ",
        Aapcs = "extern \"aapcs\" ",
        Win64 = "extern \"win64\" ",
        SysV64 = "extern \"sysv64\" ",
        System = "extern \"system\" ",
        /// In conjuction with [`Tag::UnusualAbi`].
        Other = "extern ",
    }
);

to_str!({val Tag::Decl(Decl::Struct(val))}
    pub enum Struct {
        Struct = "struct ",
    }
);

to_str!(
    /// Used to recognize the enclosing span for a component that needs styles.
    /// We want generics definitions, where bounds, names for structs/enums/fns,
    /// types for fields/enums/fn arguments etc to be colored.
    ///
    /// E.g. `Start(Function) - Path(ID) - Name(string) - End(Function)`
    ///
    /// All variants are zero str_len, and won't display anything.
    /// In conjuction with [`Tag::Start`] and [`Tag::End`].
    pub enum Span {
        PathName = "",
        TypeName = "",
        GenericsDef = "",
        WhereBound = "",
        FunctionName = "",
        FunctionArg = "",
        StructName = "",
        Field = "",
        EnumName = "",
        Variant = "",
    }
);
