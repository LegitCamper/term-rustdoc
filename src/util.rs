pub use compact_str::{format_compact as xformat, CompactString as XString, CompactStringExt};
