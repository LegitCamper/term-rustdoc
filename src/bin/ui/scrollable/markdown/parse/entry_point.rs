use super::{
    code_block, element::Element, list, segment_str, Block, Blocks, Line, LinkTag, Links, MetaTag,
    Word,
};
use pulldown_cmark::{CodeBlockKind, Event, Options, Parser, Tag, TagEnd};
use ratatui::style::{Color, Modifier, Style};
use std::ops::Range;
use term_rustdoc::util::XString;

pub fn parse(doc: &str) -> Blocks {
    if doc.is_empty() {
        return Blocks::default();
    }
    let mut blocks = Blocks::new();
    let mut iter = Parser::new_ext(
        doc,
        Options::ENABLE_FOOTNOTES
            | Options::ENABLE_STRIKETHROUGH
            | Options::ENABLE_TABLES
            | Options::ENABLE_TASKLISTS,
    )
    .into_offset_iter();
    while let Some((event, range)) = iter.by_ref().next() {
        match event {
            Event::Start(Tag::Paragraph) => {
                let mut block = Block::default();
                let mut para = ele!(iter, Paragraph, range);
                Element::new(doc, &mut block, blocks.links(), para).parse_paragraph();
                blocks.push(block);
            }
            Event::Start(Tag::CodeBlock(CodeBlockKind::Indented)) => {
                if let Some((Event::Text(code_block), _)) = iter.next() {
                    blocks.push(code_block::rust(&code_block));
                }
            }
            Event::Start(Tag::CodeBlock(CodeBlockKind::Fenced(fence))) => {
                if let Some((Event::Text(code_block), _)) = iter.next() {
                    blocks.push(code_block::parse(&mut XString::from(&*fence), &code_block));
                }
            }
            Event::Start(Tag::Heading { level, .. }) => {
                while let Some(event) = ele!(#heading iter, level, range).next() {
                    if let (Event::Text(text), _) = event {
                        blocks.push(Block::heading(level as u8, &text));
                    }
                }
            }
            Event::Rule => blocks.push(Block::from_iter([Word {
                tag: MetaTag::Rule,
                ..Default::default()
            }])),
            Event::Start(Tag::Table(_)) => {
                // the table is rendered via original contents with syntect's highlights
                blocks.push(code_block::md_table(&doc[range.clone()]));
                ele!(iter, Table, range).last(); // consume the whole table
            }
            Event::Start(Tag::BlockQuote) => {
                if let Some((Event::Start(Tag::Paragraph), range)) = iter.next() {
                    let mut block = Block::default();
                    let mut para = ele!(iter, Paragraph, range);
                    Element::new(doc, &mut block, blocks.links(), para).parse_paragraph();
                    block.set_quote_block();
                    blocks.push(block);
                }
            }
            Event::Start(Tag::FootnoteDefinition(key)) => {
                if let Some((Event::Start(Tag::Paragraph), range)) = iter.next() {
                    let mut block = Block::default();
                    let mut para = ele!(iter, Paragraph, range);
                    Element::new(doc, &mut block, blocks.links(), para).parse_paragraph();
                    block.set_foot_note();
                    blocks.links().push_footnote(&key, block);
                }
            }
            Event::Start(Tag::List(kind)) => {
                let iter = ele!(#list iter, kind.is_some(), range);
                let mut block = Block::default();
                list::parse(&mut 0, kind, iter, &mut block, doc, blocks.links());
                blocks.push(block);
            }
            _ => (),
        }
    }
    blocks.shrink_to_fit();
    blocks
}

#[test]
fn parse_markdown() {
    let doc = r#"
aaa b *c* d **e**. ~xxx~ z.

1 *c **ss** d sadsad xxx* `yyyy`

```
let a = 1;
```

> rrr sss 
> tt

- [x] done!
    - nested list
- [ ] undone
    1. *a*
    2. `b`
"#;
    parse(doc);
    insta::assert_display_snapshot!(parse(doc), @r###"
    aaa b c d e. xxx z.

    1 c ss d sadsad xxx `yyyy`

    let a = 1;

    rrr sss tt

    * [x] done!
      * nested list
    * [ ] undone
      1. a
      2. `b`

    "###);
}
