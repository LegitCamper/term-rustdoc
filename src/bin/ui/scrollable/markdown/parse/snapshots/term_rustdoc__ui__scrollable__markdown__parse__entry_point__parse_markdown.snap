---
source: src/bin/ui/scrollable/markdown/parse/entry_point.rs
expression: "markdown_iter(doc).collect::<Vec<_>>()"
---
[
    (
        Start(
            Heading {
                level: H1,
                id: None,
                classes: [],
                attrs: [],
            },
        ),
        1..6,
    ),
    (
        Text(
            Borrowed(
                "h1",
            ),
        ),
        3..5,
    ),
    (
        End(
            Heading(
                H1,
            ),
        ),
        1..6,
    ),
    (
        Start(
            Paragraph,
        ),
        7..35,
    ),
    (
        Text(
            Borrowed(
                "aaa b ",
            ),
        ),
        7..13,
    ),
    (
        Start(
            Emphasis,
        ),
        13..16,
    ),
    (
        Text(
            Borrowed(
                "c",
            ),
        ),
        14..15,
    ),
    (
        End(
            Emphasis,
        ),
        13..16,
    ),
    (
        Text(
            Borrowed(
                " d ",
            ),
        ),
        16..19,
    ),
    (
        Start(
            Strong,
        ),
        19..24,
    ),
    (
        Text(
            Borrowed(
                "e",
            ),
        ),
        21..22,
    ),
    (
        End(
            Strong,
        ),
        19..24,
    ),
    (
        Text(
            Borrowed(
                ". ",
            ),
        ),
        24..26,
    ),
    (
        Start(
            Strikethrough,
        ),
        26..31,
    ),
    (
        Text(
            Borrowed(
                "xxx",
            ),
        ),
        27..30,
    ),
    (
        End(
            Strikethrough,
        ),
        26..31,
    ),
    (
        Text(
            Borrowed(
                " z.",
            ),
        ),
        31..34,
    ),
    (
        End(
            Paragraph,
        ),
        7..35,
    ),
    (
        Start(
            Paragraph,
        ),
        36..69,
    ),
    (
        Text(
            Borrowed(
                "1 ",
            ),
        ),
        36..38,
    ),
    (
        Start(
            Emphasis,
        ),
        38..61,
    ),
    (
        Text(
            Borrowed(
                "c ",
            ),
        ),
        39..41,
    ),
    (
        Start(
            Strong,
        ),
        41..47,
    ),
    (
        Text(
            Borrowed(
                "ss",
            ),
        ),
        43..45,
    ),
    (
        End(
            Strong,
        ),
        41..47,
    ),
    (
        Text(
            Borrowed(
                " d sadsad xxx",
            ),
        ),
        47..60,
    ),
    (
        End(
            Emphasis,
        ),
        38..61,
    ),
    (
        Text(
            Borrowed(
                " ",
            ),
        ),
        61..62,
    ),
    (
        Code(
            Borrowed(
                "yyyy",
            ),
        ),
        62..68,
    ),
    (
        End(
            Paragraph,
        ),
        36..69,
    ),
    (
        Start(
            CodeBlock(
                Fenced(
                    Borrowed(
                        "",
                    ),
                ),
            ),
        ),
        70..88,
    ),
    (
        Text(
            Borrowed(
                "let a = 1;\n",
            ),
        ),
        74..85,
    ),
    (
        End(
            CodeBlock,
        ),
        70..88,
    ),
    (
        Start(
            BlockQuote,
        ),
        90..106,
    ),
    (
        Start(
            Paragraph,
        ),
        92..106,
    ),
    (
        Text(
            Borrowed(
                "rrr sss",
            ),
        ),
        92..99,
    ),
    (
        SoftBreak,
        100..101,
    ),
    (
        Text(
            Borrowed(
                "tt",
            ),
        ),
        103..105,
    ),
    (
        End(
            Paragraph,
        ),
        92..106,
    ),
    (
        End(
            BlockQuote,
        ),
        90..106,
    ),
    (
        Start(
            List(
                None,
            ),
        ),
        107..172,
    ),
    (
        Start(
            Item,
        ),
        107..137,
    ),
    (
        TaskListMarker(
            true,
        ),
        109..112,
    ),
    (
        Text(
            Borrowed(
                "done!",
            ),
        ),
        113..118,
    ),
    (
        Start(
            List(
                None,
            ),
        ),
        121..137,
    ),
    (
        Start(
            Item,
        ),
        121..137,
    ),
    (
        Text(
            Borrowed(
                "nested list",
            ),
        ),
        125..136,
    ),
    (
        End(
            Item,
        ),
        121..137,
    ),
    (
        End(
            List(
                false,
            ),
        ),
        121..137,
    ),
    (
        End(
            Item,
        ),
        107..137,
    ),
    (
        Start(
            Item,
        ),
        137..172,
    ),
    (
        TaskListMarker(
            false,
        ),
        139..142,
    ),
    (
        Text(
            Borrowed(
                "undone",
            ),
        ),
        143..149,
    ),
    (
        Start(
            List(
                Some(
                    1,
                ),
            ),
        ),
        152..172,
    ),
    (
        Start(
            Item,
        ),
        152..161,
    ),
    (
        Start(
            Emphasis,
        ),
        157..160,
    ),
    (
        Text(
            Borrowed(
                "a",
            ),
        ),
        158..159,
    ),
    (
        End(
            Emphasis,
        ),
        157..160,
    ),
    (
        End(
            Item,
        ),
        152..161,
    ),
    (
        Start(
            Item,
        ),
        163..172,
    ),
    (
        Code(
            Borrowed(
                "b",
            ),
        ),
        168..171,
    ),
    (
        End(
            Item,
        ),
        163..172,
    ),
    (
        End(
            List(
                true,
            ),
        ),
        152..172,
    ),
    (
        End(
            Item,
        ),
        137..172,
    ),
    (
        End(
            List(
                false,
            ),
        ),
        107..172,
    ),
]
